---
title: "HK-PPI-hubs"
author: "Martin Loza"
date: "2023/10/17
---

I would like to analyze housekeeping PPI hubs

NOTES:


```{r setup}
library(dplyr)
library(ggplot2)
library(patchwork)
library(stringr)
library(GenomicRanges)
library(igraph)
library(ggraph)
library(readr)

## Global variables
in_dir  <- "~/Documents/Projects/HK_Interactions/Analysis/2023_09/2023_09_18/Results/"
out_dir <- "~/Documents/Projects/HK_Interactions/Analysis/2023_10/2023_10_17/Results_deeptools/bed/"

date = 231018
#fix length
l = 3000

#Functions
source(file = "~/Documents/Projects/HK_Interactions/Analysis/Functions.R")
source(file = "~/Documents/Projects/HK_Interactions/Analysis/Functions_Visualizations.R")

#Local functions 

GetGenerationEdges <- function(edges = NULL, init_nodes = NULL, n_generations = 2){
  
  #loop. In each generation get the edges associated to init nodes
  for(i in 2:n_generations){
    tmp_df <- edges %>% filter(gene_from %in% all_of(init_nodes) | gene_to %in% all_of(init_nodes))
    init_nodes <- union(tmp_df$gene_from, tmp_df$gene_to)
  }

  return(tmp_df)
}
```

## Load data

Load annotated interactions 

```{r}
interactions <- readRDS(paste0(in_dir,"/interactions_all_230912.Rds"))
interactions %>% filter(!duplicated(region_id)) %>% pull(n_celltypes) %>% table()
```

## Set up

I would like to select interactions only for CP 

```{r}
interactions_cp <- interactions %>% filter(ann_v1 == "CP")
interactions_cp$ann_v1 <- droplevels(interactions_cp$ann_v1)
interactions_cp %>% filter(!duplicated(region_id)) %>% pull(n_celltypes) %>% table()
```

Check for duplicated interactions_cp 

```{r}
any(duplicated(interactions_cp$id))
```

Set the absolute distance of interactions_cp 

```{r}
interactions_cp <- interactions_cp %>%
  mutate(abs_distance = abs(distance))
```

## Cell type-specific CP

```{r}
table(interactions_cp$n_celltypes)
```
Let's set as CTS those CP found in less than 10% of the cell types 

```{r}
cts_interactions <- interactions_cp %>% filter(n_celltypes <= 5)
```

Save bed file of cts_cp with fixed length

```{r}
#setup bed file
cts_bed <- cts_interactions %>%
  filter(!duplicated(region_id)) %>% 
  select(chr, center) %>% 
  mutate(start = center - l, end = center + l)

#test. All regions has the same fixed length?
cts_bed %>% mutate(length = end - start) %>% mutate(tt = length == l*2) %>% pull(tt) %>% sum == nrow(cts_bed)
```

```{r}
cts_bed <- cts_bed %>% select(chr, start, end)

tmp_file_dir <- paste0(out_dir,"cts_cp_",l,"bp_",date,".bed")

#save bedpe
write.table(x = cts_bed, file = tmp_file_dir,
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)

rm(tmp_file_dir)
rm(cts_bed)
```

## HK-CP

### Hub HK-CP

Let's select only those interactions_cp found in more than 45 cell types

```{r}
#housekeeping interactions
hk_interactions <- interactions_cp %>% filter(n_celltypes == 50)
table(hk_interactions$test_n_celltypes)
```

Save bed file of hk_cp with fixed length

```{r}
#setup bed file
hk_bed <- hk_interactions %>% 
  filter(!duplicated(region_id)) %>% 
  select(chr, center) %>% 
  mutate(start = center - l, end = center + l)

#test. All regions has the same fixed length?
hk_bed %>% mutate(length = end - start) %>% mutate(tt = length == l*2) %>% pull(tt) %>% sum == nrow(hk_bed)
```

```{r}
hk_bed <- hk_bed %>% select(chr, start, end)

tmp_file_dir <- paste0(out_dir,"hk_cp_",l,"bp_",date,".bed")

#save bedpe
write.table(x = hk_bed, file = tmp_file_dir,
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)

rm(tmp_file_dir)
rm(hk_bed)
```

## HK-PPI

Get HK-CP with HK-PPI

```{r}
hk_ppi <- hk_interactions %>% filter(n_ct_interactions >= 45)
table(hk_ppi$test_n_celltypes)
```

Let's remove the bias towards self-interactions

```{r}
#interaction_id
hk_ppi <- hk_ppi %>% mutate(interaction_id = paste0(region_id,"_",gene))
#test
any(duplicated(hk_ppi$interaction_id))

#get self interactions for removing
rmv_interactions <- hk_ppi %>% filter(nearest_gene == gene & abs_distance <= 1000)

#remove the interactions
hk_ppi <- hk_ppi %>% filter(!interaction_id %in% rmv_interactions$interaction_id)
```

Get HK-CP with non-HK-PPI based on the region id

```{r}
other_hk_interactions <- hk_interactions %>% 
  filter(!region_id %in% hk_ppi$region_id)
```

Prepare beds

```{r}
#setup bed file
hk_ppi_bed <- hk_ppi %>% 
  filter(!duplicated(region_id)) %>%  
  select(chr, center) %>% 
  mutate(start = center - l, end = center + l)

#test. All regions has the same fixed length?
hk_ppi_bed %>% mutate(length = end - start) %>% mutate(tt = length == l*2) %>% pull(tt) %>% sum == nrow(hk_ppi_bed)

#setup bed file
hk_other_bed <- other_hk_interactions %>% 
  filter(!duplicated(region_id)) %>%  
  select(chr, center) %>% 
  mutate(start = center - l, end = center + l)

#test. All regions has the same fixed length?
hk_other_bed %>% mutate(length = end - start) %>% mutate(tt = length == l*2) %>% pull(tt) %>% sum == nrow(hk_other_bed)
```

Save beds

```{r}
hk_ppi_bed <- hk_ppi_bed %>% select(chr, start, end)
hk_other_bed <- hk_other_bed %>% select(chr, start, end)

#save bed
tmp_file_dir <- paste0(out_dir,"hk_ppi_",l,"bp_",date,".bed")
write.table(x = hk_ppi_bed, file = tmp_file_dir,
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)

tmp_file_dir <- paste0(out_dir,"hk_other_",l,"bp_",date,".bed")
write.table(x = hk_other_bed, file = tmp_file_dir,
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)

rm(tmp_file_dir)
rm(hk_ppi_bed, hk_other_bed)
```

### HK-PPI hubs

For now, let's assume that the regions corresponds to an annotated gene
Later, let's check for divergent genes too. For now, let's use the nearest genes

Before we filter out sel-interactions. Let's not do it this time. Anyways, genes with only self-interactions will not probably selected as hubs

```{r}
#this is step will focus only in HK-PPI, removing noise from nonHK-PPI
hk_ppi <- hk_interactions %>% filter(n_ct_interactions >= 45)
table(hk_ppi$test_n_celltypes)
```

We need to get a data frame with the edges and the weight information

```{r}
#set new genes with direction
hk_ppi <- hk_ppi %>% 
  mutate(gene_from = nearest_gene, gene_to = gene) %>%
  mutate(weight = n_ct_interactions/50) #normalized weights
#get the edges data frames
edges_df <- hk_ppi %>% dplyr::select(gene_from, gene_to, weight)
```

Get the directed graph

```{r}
#get graph from edges
graph_dir<- graph.data.frame(d = edges_df, directed = TRUE)
```

In the case of weighted graphs, the degree it's called strength. 
Let's get the strength of the nodes

```{r}
strength <- strength(graph = graph_dir, mode = "all", loops = TRUE)
strength <- data.frame(gene = names(strength), strength = strength)
```

Let's try to select the top nodes

```{r}
n_top = 1000
top_nodes <- strength %>% slice_max(order_by = strength, n = n_top) %>% arrange(desc(strength))
hub_genes <- top_nodes$gene
```

```{r}
# let's focus on the ePromoters, promoters with enhancer capabilities
hub_ppi <- hk_ppi %>% filter(gene_from %in% hub_genes)
```

let's check how many regions we got 

```{r}
#number of unique regions
hub_ppi %>% filter(!duplicated(region_id)) %>% nrow()
```

we have more regions than unique hubs. This is because a gene hub could have more than one associated TSS. Therefore, for each gene, let's focus on the transcript with higher number of interactions.

```{r}
#get the top TSS of each hub
hub_region <- hub_ppi %>% 
  group_by(gene_from) %>%
  count(region_id) %>%
  slice_max(n, n = 1) %>% 
  ungroup() %>% pull(region_id)

cat("Number of regions: ", length(hub_region))
```

Get hub_hk and nonhub_hk

```{r}
hub_hk <- hk_interactions %>% filter(region_id %in% hub_region)
nonhub_hk <- hk_interactions %>% filter(!region_id %in% hub_region)
```

```{r}
#test
sum(nrow(hub_hk %>% filter(!duplicated(region_id))), nrow(nonhub_hk %>% filter(!duplicated(region_id)))) == nrow(hk_interactions %>% filter(!duplicated(region_id)))
```

Prepare beds

```{r}
#setup bed file
hub_hk_bed <- hub_hk %>% 
  filter(!duplicated(region_id)) %>%  
  select(chr, center) %>% 
  mutate(start = center - l, end = center + l)

#test. All regions has the same fixed length?
hub_hk_bed %>% mutate(length = end - start) %>% mutate(tt = length == l*2) %>% pull(tt) %>% sum == nrow(hub_hk_bed)

#setup bed file
nonhub_hk_bed <- nonhub_hk %>% 
  filter(!duplicated(region_id)) %>%  
  select(chr, center) %>% 
  mutate(start = center - l, end = center + l)

#test. All regions has the same fixed length?
nonhub_hk_bed %>% mutate(length = end - start) %>% mutate(tt = length == l*2) %>% pull(tt) %>% sum == nrow(nonhub_hk_bed)
```

```{r}
#test
sum(nrow(hub_hk_bed), nrow(nonhub_hk_bed)) == nrow(hk_interactions %>% filter(!duplicated(region_id)))
```

Save beds

```{r}
hub_hk_bed <- hub_hk_bed %>% select(chr, start, end)
nonhub_hk_bed <- nonhub_hk_bed %>% select(chr, start, end)

#save bed
tmp_file_dir <- paste0(out_dir,"hub_hk_",l,"bp_",date,".bed")
write.table(x = hub_hk_bed, file = tmp_file_dir,
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)

tmp_file_dir <- paste0(out_dir,"nonhub_hk_",l,"bp_",date,".bed")
write.table(x = nonhub_hk_bed, file = tmp_file_dir,
            quote = FALSE, sep = "\t", row.names = FALSE, col.names = FALSE)

rm(tmp_file_dir)
rm(hub_hk_bed, nonhub_hk_bed)
```












